File: app/src/main/java/com/yalantis/ui/adapter/RepositoryAdapter.java
package com.yalantis.ui.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.yalantis.R;
import com.yalantis.model.Repository;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class RepositoryAdapter extends RecyclerView.Adapter<RepositoryAdapter.RepositoryViewHolder> {

    private final ItemClickListener mItemClickListener;

    private List<Repository> mRepositories = new ArrayList<>();

    public RepositoryAdapter(@Nullable ItemClickListener itemClickListener) {
        mItemClickListener = itemClickListener;
    }

    public void addRepositories(List<Repository> repositoryList) {
        mRepositories.clear();
        mRepositories.addAll(repositoryList);
        notifyDataSetChanged();
    }

    @Override
    public RepositoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_repository, parent, false);
        return new RepositoryViewHolder(itemView, mItemClickListener);
    }

    @Override
    public void onBindViewHolder(RepositoryViewHolder holder, int position) {
        Repository repository = mRepositories.get(position);
        holder.bindData(repository);
    }

    @Override
    public int getItemCount() {
        return mRepositories.size();
    }

    public static class RepositoryViewHolder extends RecyclerView.ViewHolder {

        @Bind(R.id.text_view_title)
        TextView titleTextView;
        @Bind(R.id.text_view_description)
        TextView descriptionTextView;

        private final ItemClickListener mItemClickListener;

        private Repository mRepository;

        public RepositoryViewHolder(@NonNull View itemView, @Nullable ItemClickListener itemClickListener) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            mItemClickListener = itemClickListener;
        }

        public void bindData(Repository repository) {
            mRepository = repository;

            titleTextView.setText(repository.getName());
            descriptionTextView.setText(repository.getDescription());
        }

        @OnClick(R.id.linear_layout_content)
        void onClickListItem() {
            if (mItemClickListener != null) {
                mItemClickListener.onItemClick(mRepository);
            }
        }

    }

    public interface ItemClickListener {
        void onItemClick(Repository repository);
    }

}

File: app/src/main/java/com/yalantis/flow/repository/RepositoryAdapter.kt
package com.yalantis.flow.repository

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.yalantis.R
import com.yalantis.data.Repository
import java.util.*

internal class RepositoryAdapter(val onClick: (Repository) -> Unit) : RecyclerView.Adapter<RepositoryAdapter.RepositoryViewHolder>() {

    private val mRepositories = ArrayList<Repository>()

    fun addRepositories(repositoryList: List<Repository>) {
        mRepositories.clear()
        mRepositories.addAll(repositoryList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RepositoryViewHolder {
        val binding: View = LayoutInflater.from(parent.context).inflate(R.layout.item_repository, parent, false)
        return RepositoryViewHolder(binding, onClick)
    }

    override fun onBindViewHolder(holder: RepositoryViewHolder, position: Int) {
        val repository = mRepositories[position]
    }

    override fun getItemCount(): Int {
        return mRepositories.size
    }

    internal class RepositoryViewHolder(val binding: View, val onClick: (Repository) -> Unit) : RecyclerView.ViewHolder(binding) {

        fun bindData(repository: Repository) {
            with(repository) {
//                text_view_title.text = name
//                text_view_description.text = description
                binding.setOnClickListener { onClick(this) }
            }
        }

    }
}